File: 백준/Gold/1135. 뉴스 전하기/뉴스 전하기.java
import java.io.*;
import java.util.*;

public class Main {
    static int N = 0;
    static List<List<Integer>> arr = new ArrayList<>();

    static int solve(int idx) {
        int result = 0;

        int sz = arr.get(idx).size();
        if(sz == 0) return 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for(final int c : arr.get(idx)) {
            pq.offer(solve(c));
        }

        for(int i=0; i<sz; ++i) {
            result = Math.max(result, pq.poll() + i + 1);
        }

        return result;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int answer = 0;
        N = Integer.parseInt(br.readLine());
        for(int i=0; i<N; ++i)
            arr.add(new ArrayList<>());
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int inp = Integer.parseInt(st.nextToken());
        for(int i=1; i<N; ++i) {
            inp = Integer.parseInt(st.nextToken());
            arr.get(inp).add(i);
        }

        answer = solve(0);
        System.out.print(answer);
    }
}

File: 백준/Gold/1135. 뉴스 전하기/뉴스 전하기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
vector<vector<int>> children;
int solve(const int idx)
{
    int sz = children[idx].size();
    if(sz == 0) return 0;

    int result = 0;
    vector<int> results;
    for (const auto& c : children[idx])
        results.push_back(solve(c));

    std::sort(results.rbegin(), results.rend());

    for (int order = 0; order < sz; ++order) 
        result = max(result, results[order] + order + 1);

    return result;
}

int main() {

    int N, inp;
    cin >> N >> inp;
    
    children.resize(N);
    for(int i=1; i<N; ++i) {
        cin >> inp;
        children[inp].push_back(i);
    }

    cout << solve(0);
    return 0;
}